eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -0,0 +1,347 @@
+      }
+      case WrappedMultiPartPaymentFailed(MultiPartPaymentFSM.MultiPartPaymentFailed(_, failure, parts)) =>
+        context.log.warn("could not relay payment (paidAmount={} failure={})", parts.map(_.amount).sum, failure)
+        Metrics.recordPaymentRelayFailed(failure.getClass.getSimpleName, Tags.RelayType.Trampoline)
+        handler ! PoisonPill
+        parts.collect { case p: MultiPartPaymentFSM.HtlcPart => rejectHtlc(p.htlc.id, p.htlc.channelId, p.amount, Some(failure)) }
+        Behaviors.stopped
+      case WrappedMultiPartPaymentSucceeded(MultiPartPaymentFSM.MultiPartPaymentSucceeded(_, parts)) =>
+        val upstream = Upstream.Trampoline(htlcs)
+        handler ! PoisonPill
+        validateRelay(nodeParams, upstream, nextPayload) match {
+          case Some(failure) =>
+            context.log.warn(s"rejecting trampoline payment (amountIn=${upstream.amountIn} expiryIn=${upstream.expiryIn} amountOut=${nextPayload.amountToForward} expiryOut=${nextPayload.outgoingCltv} htlcCount=${parts.length} reason=$failure)")
+            rejectPayment(upstream, Some(failure))
+            Behaviors.stopped
+          case None =>
+            context.log.info(s"relaying trampoline payment (amountIn=${upstream.amountIn} expiryIn=${upstream.expiryIn} amountOut=${nextPayload.amountToForward} expiryOut=${nextPayload.outgoingCltv} htlcCount=${upstream.adds.length})")
+            val payFSM = relay(upstream, nextPayload, nextPacket)
+            sending(payFSM, upstream, nextPayload, fulfilledUpstream = false)
+        }
+      case _ => Behaviors.unhandled
+    }
+
+  /**
+   * Once the payment is forwarded, we're waiting for fail/fulfill responses from downstream nodes.
+   *
+   * @param upstream          complete HTLC set received.
+   * @param nextPayload       relay instructions.
+   * @param fulfilledUpstream true if we already fulfilled the payment upstream.
+   */
+  private def sending(payFSM: ActorRef, upstream: Upstream.Trampoline, nextPayload: Onion.NodeRelayPayload, fulfilledUpstream: Boolean): Behavior[Command] =
+    Behaviors.receiveMessage {
+      rejectExtraHtlcPartialFunction orElse {
+        // this is the fulfill that arrives from downstream channels
+        case WrappedPreimageReceived(PreimageReceived(_, paymentPreimage)) =>
+          if (!fulfilledUpstream) {
+            // We want to fulfill upstream as soon as we receive the preimage (even if not all HTLCs have fulfilled downstream).
+            context.log.debug("trampoline payment successfully relayed")
+            fulfillPayment(upstream, paymentPreimage)
+            sending(payFSM, upstream, nextPayload, fulfilledUpstream = true)
+          } else {
+            // we don't want to fulfill multiple times
+            Behaviors.same
+          }
+        case WrappedPaymentSent(paymentSent) =>
+          // We may have already fulfilled upstream, but we can now emit an accurate relayed event and clean-up resources.
```

Comment:
nit: this comment should go in the `success` handler

Reply:
Fixed with 5467945e8.